test/unit/io_managers_test.py

```diff
@@ -249,3 +255,232 @@ def test_error_when_reading_view_without_metadata(pudl_sqlite_io_manager_fixture
+                "date": datetime.date(2021, 12, 31),
+                "publication_time": datetime.datetime(2022, 2, 1, 0, 0, 0),
+                "str_factoid": "original 2021 EOY value",
+            },
+            {
+                "entity_id": "C000001",
+                "utility_type_axis": "electric",
+                "filing_name": "Utility_Co_0002",
+                "date": datetime.date(2021, 12, 31),
+                "publication_time": datetime.datetime(2022, 2, 1, 1, 1, 1),
+                "str_factoid": "updated 2021 EOY value",
+            },
+        ]
+    )
+
+    id_table = pd.DataFrame.from_records(
+        [
+            {"filing_name": "Utility_Co_0001", "report_year": 2021},
+            {"filing_name": "Utility_Co_0002", "report_year": 2021},
+        ]
+    )
+
+    conn = sa.create_engine(f"sqlite:///{db_path}")
+    df.to_sql("test_table_instant", conn)
+    id_table.to_sql("identification_001_duration", conn)
+
+    input_context = build_input_context(
+        asset_key=AssetKey("test_table_instant"),
+        resources={
+            "dataset_settings": mocker.MagicMock(
+                ferc1=mocker.MagicMock(xbrl_years=[2021])
+            )
+        },
+    )
+    io_manager = FercXBRLSQLiteIOManager(base_dir=tmp_path, db_name="test_db")
+    observed_table = io_manager.load_input(input_context)
+
+    assert len(observed_table) == 1
+    assert observed_table.str_factoid.to_numpy().item() == "updated 2021 EOY value"
+
+
+example_schema = pa.DataFrameSchema(
+    {
+        "entity_id": pa.Column(str, nullable=False),
+        "date": pa.Column("datetime64[ns]", nullable=False),
+        "utility_type": pa.Column(
+            str, pa.Check.isin(["electric", "gas", "total", "other"]), nullable=False
+        ),
+        "publication_time": pa.Column("datetime64[ns]", nullable=False),
+        "int_factoid": pa.Column(int),
+        "float_factoid": pa.Column(float),
+        "str_factoid": pa.Column("str"),
+    }
+)
+
+
+@hypothesis.given(example_schema.strategy(size=3))
+def test_get_unique_row_per_context(df):
+    context_cols = ["entity_id", "date", "utility_type"]
+    deduped = FercXBRLSQLiteIOManager.use_latest_filing_for_context(df, context_cols)
+    example_schema.validate(deduped)
+
+    # every post-deduplication row exists in the original rows
+    assert (deduped.merge(df, how="left", indicator=True)._merge != "left_only").all()
+    # for every [entity_id, utility_type, date] - th"true"e is only one row
+    assert (~deduped.duplicated(subset=context_cols)).all()
+    # for every *context* in the input there is a corresponding row in the output
+    original_contexts = df.groupby(context_cols, as_index=False).last()
+    paired_by_context = original_contexts.merge(
+        deduped, on=context_cols, how="outer", suffixes=["_in", "_out"], indicator=True
+    ).set_index(context_cols)
+    hypothesis.note(original_contexts)
+    hypothesis.note(deduped)
```

Comment:
What does `hypothesis.note()` do? I read the docs but I'm still confused.

> The note is printed for the minimal failing example of the test in order to include any additional information you might need in your test.

Reply:
Hypothesis runs a lot of iterations of each test, right - and once there's a failure, they try to find the simplest failing input.

So if you were to log things out, you'd have a lot of output to sift through. And if you were to stick a debugger in the test, you'd end up hitting the breakpoint hundreds of times.

`note()` is like `print()` but instead of printing *every* time, it prints output *only* for the minimal failing example, so you can understand what's going on when it fails.

Reply:
Ah okay that's really cool!

Reply:
So if `hypothesis.note()` results in output, that means there's a failure, so if all is well we shouldn't be seeing any output.